### DISCUSSION TITLE
Error: null value in column "id" of relation "my_items" violates not-null constraint

### DISCUSSION THREAD

**Discussion by WhereJuly** [mikro-orm/mikro-orm#5768] [Q&A] [answered]
I get the Postgres error when I update a collection. It appears only on stage environment, I cannot catch it onto my development environment however I try. The development and stage environments assumed to be identical. Databases are `schema:fresh`'ed on both, the schema definitions and the entire code is just a repository clone.

The only difference I can see is the Mikro-ORM debug output for the SQL statement in question between local and development environments.

On development it says: 
``` sql
insert into "my_items" ("created_at", "updated_at"... 
values 
('2024-07-02T15:06:10.339Z', '2024-07-02T15:06:10.339Z'...)
('2024-07-02T15:06:10.339Z', '2024-07-02T15:06:10.339Z'...)
``` 
that executes successfully.

On stage it says:
``` sql
insert into "my_items" ("id", "created_at", "updated_at", ... 
values 
(7, '2024-07-02T13:46:37.264Z', '2024-07-02T13:46:37.264Z'...)
(NULL, '2024-07-02T13:46:37.264Z', '2024-07-02T13:46:37.264Z'...)
```
That throws the subject error with detail
`Failing row contains (null, 2024-07-02 13:46:37.264+00, 2024-07-02 13:46:37.264+00...`  
so that on stage environment the `id` mentioned in `insert` statement paired with `NULL` in the row data produces the error.

The strange thing for me is `id` is not provided in local SQL statement while it is in stage's one.

Why would Mikro-ORM put or do not put ID into the statement for the same entity and operation? Appreciate any advice.

**Comment by B4nan** [comment]
The value is not coming out of nowhere, it needs to be provided explicitly (by your code). Hard to guess without seeing any code.

**Reply by WhereJuly** [reply]
> The value is not coming out of nowhere, it needs to be provided explicitly (by your code). Hard to guess without seeing any code.

@B4nan Thanks. I replicated the error on local environment running both front and backend here. The `id` error did not appear in local tests, so will search how test request payload differs from frontend payload.

**Reply by WhereJuly** [reply, marked as answer]
@B4nan the culprit was the collection under update contained 2 items. The one existing with an actual ID and the other new one with ID=null.

Thus Mikro-ORM based on first item content decided to add `ID` into statement, that contradicted `not null` constraint for ID field and the second item triggered the error.

Preparing the data for `em.create(entity)` I forcefully assigned `entity.id = undefined` and that fixed the issue.

Thanks again for a quick reaction and valuable heads-up.
